share/availability/discovery/discovery.go

```diff
@@ -183,23 +199,42 @@ func (d *Discovery) ensurePeers(ctx context.Context) {
 			log.Info("Context canceled. Finishing peer discovery")
 			return
 		case <-t.C:
-			if uint(d.set.Size()) == d.peersLimit {
-				// stop ticker if we have reached the limit
-				t.Stop()
-				continue
-			}
-			peers, err := d.disc.FindPeers(ctx, topic)
-			if err != nil {
-				log.Error(err)
-				continue
-			}
-			for p := range peers {
-				go d.handlePeerFound(ctx, topic, p)
-			}
+			d.findPeers(ctx)
 		}
 	}
 }
 
+func (d *Discovery) findPeers(ctx context.Context) {
+	if uint(d.set.Size()) >= d.peersLimit {
+		log.Debugw("at peer limit, skipping FindPeers", "size", d.set.Size())
+		return
+	}
+
+	peers, err := d.disc.FindPeers(ctx, topic)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	// we use a wait group to avoid a context leak
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(ctx)
+
+	for p := range peers {
+		wg.Add(1)
+		go func(peer peer.AddrInfo) {
+			defer wg.Done()
+			d.handlePeerFound(ctx, peer, cancel)
+		}(p)
+	}
+
+	// Wait for all goroutines to finish, then cancel the context
+	go func() {
+		wg.Wait()
+		cancel()
+	}()
```

Comment:
I wondert why this would be necessary. This logic makes cancel only after the FindPeers query is finished(peers chan closed) and all the peers are handled with any result(wg.Wait). Nothing is technically left to cancel in this case, as everything is finished.

I would keep the wg and make the `findPeers` block on it so that we don't start new `findPeers` before the previous one hasn't finished yet. Also, I would remove the cancel after wg, as it effectively noop. The only issue is that wg.Wait does not respect the context, but this can be fixed with errgroup

Reply:
The only reason the waitgroup exists is to avoid a context leak. We leak context in case the amount of peers returned from FindPeers doesn't result in a number of connections >= peerLimit

Reply:
I don't get the leak either. By the time wg.Wait() returns, there are no any other `d.handlePeerFound` call running in goroutines.

Reply:
I will show an alternative soon in this branch